WIP: Upgrade to Hyper v0.12

TheBiggerGuy opened the pull request:
A first **WIP** version for a upgrade to Hyper v0.12. I am throwing this over the wall to get some feedback.

TheBiggerGuy committed:
Upgrade to Hyper v0.12

This upgrade Hyper to v0.12 and updats to code to work for it. It has
being done with the minimum code change and so the logic is still
aukward for the futures model. This should be addressed in later commits
but I did not want to compilcate an already large commit.

TheBiggerGuy commented on src/device.rs:
```diff
@@ -46,7 +46,7 @@ impl<C> Flow for DeviceFlow<C> {
 }
 impl<C> DeviceFlow<C>
-    where C: BorrowMut<hyper::Client>
+    where C: BorrowMut<hyper::Client<hyper::client::HttpConnector, hyper::Body>>
```
I think it would be nice to alias this type. Maybe ```type OauthHyperClient = hyper::Client<hyper::client::HttpConnector, hyper::Body>```.

Also this needs to be moved to the Http**s** type.

TheBiggerGuy commented on Cargo.toml:
```diff
@@ -13,7 +13,8 @@ edition = "2018"
 chrono = "0.4"
-hyper = "0.10.2"
+hyper = "0.12"
+hyper-tls = "0.3"
```
What Hyper TLS/HTTPS lib is the project heading towards?

dermesser replied:
to be honest, I've lost orientation on which there are. Are you favoring one?

TheBiggerGuy commented on src/authenticator.rs:
```diff
@@ -335,15 +345,19 @@ mod tests {
-                        hyper::Client::with_connector(<MockGoogleAuth as Default>::default()),
-                        <MemoryStorage as Default>::default(), None)
-                        .token(&["https://www.googleapis.com/auth/youtube.upload"]);
+        hyper::rt::run(hyper::rt::lazy(|| {
```
We need to embed (with option of overriding) a runtime inside the Authenticator so we do  not require the calling application to be inside a Tokio runtime.

In general this move by Hyper to require all applications to have there main inside a Tokio future with no return type seems like a bad design. I am sure I must be missing something. I hope to embed a runtime side Authenticator and then use that.

dermesser commented:
I just took a general look at it, and it seems like it's close to what I've been wanting to do for some time. However, the reason I haven't is compatibility with https://github.com/Byron/google-apis-rs (which is the main user of this crate so far). I see two ways forward: either update the google-apis-rs crate (which probably would make many people happy, but looked to be a lot of work every time I've looked so far), or develop this change on a branch and publish a new major version, pinning google-apis-rs on the "legacy" one.

The reason for this is that yup-oauth2 and google-apis-rs share historical roots, so I wouldn't like to break them just like that. Let me know which option you prefer: I think a new branch would help the most. In any case, I'm really happy about this!

TheBiggerGuy committed:
Update version from 1.x to 2.x

Update version number and docs as the move to Hyper v0.12 will cause a
breaking change to consumers of this lib.

TheBiggerGuy commented:
My reason for working on the Hyper update was due to my use of `google-apis-rs`. My thoughts where to make a major version change in `yup-oauth2` and let `google-apis-rs` stay on the 1.x branch until it is updated. The only fix needed in `google-apis-rs` is https://github.com/Byron/google-apis-rs/pull/219 . This starts the chicken and egg question and I do not believe ether project will be long term sustainable sticking with the dead Hyper v0.11 branch.

I have updated the pull request with the version bump. I am now 'near happy' with the code. I am unsure of the best Hyper Client API. It seems you are required to provide an executor to Hyper when not running in a `tokio::run` context.
```rust
let https = HttpsConnector::new(4).unwrap();
let runtime = tokio::runtime::Runtime::new().unwrap();
let client = hyper::Client::builder()
            .executor(runtime.executor())
            .build(https);
```
This just seems overly complicated to push on the crates consumers.

dermesser commented:
I agree with your intentions about branching and your assessment about long term viability and usage of hyper versions.

W.r.t. the tokio boilerplate, is it feasible to provide a utility function for non-tokio users?

iotaaxel commented:
- Like @TheBiggerGuy, I also need the `google-apis-rs` to update to hyper 0.12. 
- I agree with you that this change can be merged instead to another branch. 

**Initial thoughts:** 
- After that, another branch (let's call this `alternative-branch`) can even be made for `google-apis-rs` to do [Byron/google-apis-rs#219  ](Byron/google-apis-rs#219)
- Once that is done, people can be refer to `alternative-branch` if they want to use hyper 0.12.
- After that, if it is thought of as necessary, both crates can integrate the changes needed to use hyper 0.12 into the `master` branch.

iotaaxel commented:
What's the status on this?

mitsuhiko commented:
@dermesser I think at this point publishing a 2.0.0 of this library is not a terrible idea. google-apis-rs can't really upgrade until this is upgraded anyways.

dermesser commented:
I didn't check the details; could you resolve the conflicts displayed here? Otherwise I'm very happy to merge and publish.

dermesser commented:
Nevermind, I found a minute to merge it :)

dermesser commented:
Thank you so much for this PR, the work is much appreciated :) I personally find it hard sometimes to keep up with the speed of Rust and its libraries' development, and I'm sorry for this having taken so long.

dermesser commented:
I uploaded 2.0.0 now.
